### PR TITLE
css.experimental.customData

### PR DESCRIPTION
**Pull request by octref**
For #64164.

Related: https://github.com/Microsoft/vscode-css-languageservice/pull/142.

### CODE REVIEW

File: extensions/css-language-features/client/src/cssMain.ts [hunk 1]
```diff
@@ -30,13 +30,30 @@ export function activate(context: ExtensionContext) {
 
 	let documentSelector = ['css', 'scss', 'less'];
 
+	let dataPaths: string[] = workspace.getConfiguration('css').get('experimental.customData', []);
```
Comment (aeschli) [inline extensions/css-language-features/client/src/cssMain.ts]: I suspect this is just for teting purposes, but I assume in the end, these files will not be located in the workspace but part of a extension, correct?

File: extensions/css-language-features/server/src/cssServerMain.ts [hunk 2]
```diff
@@ -50,6 +52,8 @@ let scopedSettingsSupport = false;
 let foldingRangeLimit = Number.MAX_VALUE;
 let workspaceFolders: WorkspaceFolder[];
 
+const languageServices: { [id: string]: LanguageService } = {};
```
Comment (aeschli) [inline extensions/css-language-features/server/src/cssServerMain.ts]: I guess that's not really a necessary change.

File: extensions/css-language-features/server/src/languageFacts.ts [hunk 3]
```diff
@@ -0,0 +1,21 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+export function parseCSSData(source: string) {
```
Comment (aeschli) [inline extensions/css-language-features/server/src/languageFacts.ts]: I'd suggest to move that to `util`

File: extensions/css-language-features/client/src/cssMain.ts (hunk 1 above)
Comment (octref) [reply extensions/css-language-features/client/src/cssMain.ts]: They'll, see an example for HTML: https://github.com/octref/web-components-examples

Don't think we can assume everyone wants to publish extensions for enhancing their CSS.

File: extensions/css-language-features/server/src/cssServerMain.ts (hunk 2 above)
Comment (octref) [reply extensions/css-language-features/server/src/cssServerMain.ts]: Doing this so I can locally scope the data variables. I don't want the `customData` to be on top level.

File: extensions/css-language-features/client/src/cssMain.ts [hunk 4]
```diff
@@ -30,13 +30,30 @@ export function activate(context: ExtensionContext) {
...
+	if (dataPaths && dataPaths.length > 0) {
+		if (!workspace.workspaceFolders) {
+			dataPaths = [];
+		} else {
+			try {
+				const workspaceRoot = workspace.workspaceFolders[0].uri.fsPath;
```
[the remark this answers is not in the document]
Comment (octref) [reply extensions/css-language-features/client/src/cssMain.ts]: Yeah, I need to handle that. Have never tried custom filesystem providers though. Tracked in https://github.com/Microsoft/vscode/issues/66307.

**Comment by Bradleyzgoodstuff** [comment]
JamesBaker esk. Notebooks full of code a run run run till fail automatic fix boot thing R.I.P.jimmyboy

**Comment by Bradleyzgoodstuff** [comment]
Foward up into

File: extensions/css-language-features/server/src/cssServerMain.ts (hunk 2 above)
Comment (minkir014) [reply extensions/css-language-features/server/src/cssServerMain.ts]: So you say vs code can complete css variables by setting this setting

**Pull request merged**